staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/skipnonapplied.go

```diff
@@ -28,25 +30,48 @@ type skipNonAppliedManager struct {
 	objectCreater          runtime.ObjectCreater
 	gvk                    schema.GroupVersionKind
 	beforeApplyManagerName string
+	probability            float32
 }
 
 var _ Manager = &skipNonAppliedManager{}
 
 // NewSkipNonAppliedManager creates a new wrapped FieldManager that only starts tracking managers after the first apply.
 func NewSkipNonAppliedManager(fieldManager Manager, objectCreater runtime.ObjectCreater, gvk schema.GroupVersionKind) Manager {
+	return NewProbabilisticSkipNonAppliedManager(fieldManager, objectCreater, gvk, 0.0)
+}
+
+// NewProbabilisticSkipNonAppliedManager creates a new wrapped FieldManager that starts tracking managers after the first apply,
+// or starts tracking on create with p probability.
+func NewProbabilisticSkipNonAppliedManager(fieldManager Manager, objectCreater runtime.ObjectCreater, gvk schema.GroupVersionKind, p float32) Manager {
 	return &skipNonAppliedManager{
 		fieldManager:           fieldManager,
 		objectCreater:          objectCreater,
 		gvk:                    gvk,
 		beforeApplyManagerName: "before-first-apply",
+		probability:            p,
 	}
 }
 
 // Update implements Manager.
 func (f *skipNonAppliedManager) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
-	if len(managed.Fields()) == 0 {
+	accessor, err := meta.Accessor(liveObj)
+	if err != nil {
 		return newObj, managed, nil
 	}
+
+	// If managed fields is empty, we need to determine whether to skip tracking managed fields.
+	if len(managed.Fields()) == 0 {
+		// Check if the operation is a create, by checking whether lastObj's UID is empty.
+		// If the operation is create, P(tracking managed fields) = f.probability
+		// If the operation is update, skip tracking managed fields, since we already know managed fields is empty.
+		if len(accessor.GetUID()) == 0 {
+			if f.probability <= rand.Float32() {
```

Comment:
I have a preference for computing a number by e.g. hashing namespace and name so that it'll be repeatable based on the object. (e.g., if we break an e2e test, let's try to always break the same one)

Reply:
That sounds reasonable to me, at least if we are planning to get this to 100% before 1.18. 

Do you think we want this probability to be configurable by end users though?

Reply:
I don't think that's necessary, and we can always add that if it looks like we won't get to 100% by the release day.

Reply:
> if we break an e2e test, let's try to always break the same one

I'm pretty sure e2e tests generate a namespace with a randomized suffix, fyi

Reply:
That's true, and makes this less important. I still think it'd be better if it were deterministic :/